runtime/jailer.go

```diff
@@ -0,0 +1,476 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+)
+
+const (
+	jailerFolder          = "jail"
+	firecrackerBinName    = "firecracker"
+	kernelImageFileName   = "kernel-image"
+	jailerHandlerName     = "firecracker-containerd-jail-handler"
+	jailerFifoHandlerName = "firecracker-containerd-jail-fifo-handler"
+	runcConfigPath        = "/etc/containerd/firecracker-runc-config.json"
+	rootfsFolder          = "rootfs"
+
+	defaultMinIDCount = 1000
```

Comment:
If we are going to use dynamic uids/gids, I think we should start much higher than 1000 by default. In my experience, 1000 is a common starting place for uids on various distros. Even though we check that the uid/gid is not registered in `/etc/passwd` we can't prevent race conditions where a new user is registered after we do the check, so trying to stay far away from any commonly used range is preferable.

[Systemd uses 61184-65519](http://0pointer.net/blog/dynamic-users-with-systemd.html), we should probably also try to not conflict with that range.